srml/grandpa/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+		PendingChange get(pending_change): Option<StoredPendingChange<T::BlockNumber, T::SessionKey>>;
+	}
+	add_extra_genesis {
+		config(authorities): Vec<(T::SessionKey, u64)>;
+
+		build(|storage: &mut primitives::StorageMap, _: &mut primitives::ChildrenStorageMap, config: &GenesisConfig<T>| {
+			use codec::{Encode, KeyedVec};
+
+			let auth_count = config.authorities.len() as u32;
+			config.authorities.iter().enumerate().for_each(|(i, v)| {
+				storage.insert((i as u32).to_keyed_vec(
+					::fg_primitives::well_known_keys::AUTHORITY_PREFIX),
+					v.encode()
+				);
+			});
+			storage.insert(
+				::fg_primitives::well_known_keys::AUTHORITY_COUNT.to_vec(),
+				auth_count.encode(),
+			);
+		});
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		/// Report some misbehaviour.
+		fn report_misbehavior(origin, _report: Vec<u8>) -> Result {
+			ensure_signed(origin)?;
+			// TODO: https://github.com/paritytech/substrate/issues/1112
+			Ok(())
+		}
+
+		fn on_finalise(block_number: T::BlockNumber) {
+			if let Some(pending_change) = <PendingChange<T>>::get() {
+				if block_number == pending_change.scheduled_at {
+					Self::deposit_log(RawLog::AuthoritiesChangeSignal(
+						pending_change.delay,
+						pending_change.next_authorities.clone(),
+					));
+				}
+
+				if block_number == pending_change.scheduled_at + pending_change.delay {
+					Self::deposit_event(
+						RawEvent::NewAuthorities(pending_change.next_authorities.clone())
+					);
+					<AuthorityStorageVec<T::SessionKey>>::set_items(pending_change.next_authorities);
+					<PendingChange<T>>::kill();
+				}
+			}
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Get the current set of authorities, along with their respective weights.
+	pub fn grandpa_authorities() -> Vec<(T::SessionKey, u64)> {
+		<AuthorityStorageVec<T::SessionKey>>::items()
+	}
+
+	/// Schedule a change in the authorities.
+	///
+	/// The change will be applied at the end of execution of the block
+	/// `in_blocks` after the current block. This value may be 0, in which
+	/// case the change is applied at the end of the current block.
+	///
+	/// No change should be signalled while any change is pending. Returns
+	/// an error if a change is already pending.
+	pub fn schedule_change(
+		next_authorities: Vec<(T::SessionKey, u64)>,
+		in_blocks: T::BlockNumber,
+	) -> Result {
+		if Self::pending_change().is_none() {
+			let scheduled_at = system::ChainContext::<T>::default().current_height();
+			<PendingChange<T>>::put(StoredPendingChange {
+				delay: in_blocks,
+				scheduled_at,
+				next_authorities,
+			});
+
+			Ok(())
+		} else {
+			Err("Attempt to signal GRANDPA change with one already pending.")
+		}
+	}
+
+	/// Deposit one of this module's logs.
+	fn deposit_log(log: Log<T>) {
+		<system::Module<T>>::deposit_log(<T as Trait>::Log::from(log).into());
+	}
+}
+
+impl<T: Trait> Module<T>
```

Comment:
style: would join these two lines (follow on lines should always be an indent level more)